tutorials/04-hidden-markov-model/index.qmd

```diff
@@ -190,4 +195,85 @@ stationary. We can use the diagnostic functions provided by [MCMCChains](https:/
+
+## Efficient Inference With The Forward Algorithm
+
+While the above method works well for the simple example in this tutorial, some users may desire a more efficient method, especially when their model is more complicated. One simple way to improve inference is to marginalize out the hidden states of the model with an appropriate algorithm, calculating only the posterior over the continuous random variables. Not only does this allow more efficient inference via Rao-Blackwellization, but now we can sample our model with `NUTS()` alone, which is usually a much more performant MCMC kernel.
+
+Thankfully, [HiddenMarkovModels.jl](https://github.com/gdalle/HiddenMarkovModels.jl) provides an extremely efficient implementation of many algorithms related to Hidden Markov Models. This allows us to re-write our model as:
+
+```{julia}
+#| output: false
+using HiddenMarkovModels
+using FillArrays
+using LinearAlgebra
+using LogExpFunctions
+
+
+@model function BayesHmm2(y, K)
+    m ~ Bijectors.ordered(MvNormal([1.0, 2.0, 3.0], 0.5I))
+    T ~ filldist(Dirichlet(Fill(1/K, K)), K)
+
+    hmm = HMM(softmax(ones(K)), copy(T'), [Normal(m[i], 0.1) for i in 1:K])
+    Turing.@addlogprob! logdensityof(hmm, y)
+end
+
+chn2 = sample(BayesHmm2(y, 3), NUTS(), 1000)
+```
+
+
+We can compare the chains of these two models, confirming the posterior estimate is similar (modulo label switching concerns with the Gibbs model):
+```{julia}
+#| code-fold: true
+#| code-summary: "Plotting Chains"
+
+plot(chn["m[1]"], label = "m[1], Model 1, Gibbs", color = :lightblue)
+plot!(chn2["m[1]"], label = "m[1], Model 2, NUTS", color = :blue)
+plot!(chn["m[2]"], label = "m[2], Model 1, Gibbs", color = :pink)
+plot!(chn2["m[2]"], label = "m[2], Model 2, NUTS", color = :red)
+plot!(chn["m[3]"], label = "m[3], Model 1, Gibbs", color = :yellow)
+plot!(chn2["m[3]"], label = "m[3], Model 2, NUTS", color = :orange)
+```
+
+
+### Recovering Marginalized Trajectories
+
+We can use the `viterbi()` algorithm, also from the `HiddenMarkovModels` package, to recover the most probable state for each parameter set in our posterior sample:
+```{julia}
+#| output: false
+@model function BayesHmmRecover(y, K, IncludeGenerated = false)
+
+    m ~ Bijectors.ordered(MvNormal([1.0, 2.0, 3.0], 0.5I))
+    T ~ filldist(Dirichlet(Fill(1/K, K)), K)
+
+    hmm = HMM(softmax(ones(K)), copy(T'), [Normal(m[i], 0.1) for i in 1:K])
+    Turing.@addlogprob! logdensityof(hmm, y)
+
+    # Conditional generation of the hidden states.
+    if IncludeGenerated
+        seq, _ = viterbi(hmm, y)
+        s := [m[s] for s in seq]
+    else
+        return nothing
+    end
+end
+
+chn_recover = sample(BayesHmmRecover(y, 3, true), NUTS(), 1000)
+```
+
+Plotting the estimated states, we can see that the results align well with our expectations:
+
+```{julia}
+#| code-fold: true
+#| code-summary: "HMM Plotting Functions"
+
+p = plot(xlim=(0, 30), ylim=(-1, 5), size=(500, 250))

```

Comment:
I really don't like how this vector — which contains no information not present in the plot, and which is not part of anyones actual workflow (in practice, people just import data via csvs etc.) — takes up so much space on the page. 

that is why I originally introduced the folder code blocks (I agree that the other use especially is excessive, but I am more torn here).

I am considering doing something like

```
y = [i for i in vcat(fill(1.0, 6), fill(2.0, 6), fill(3.0, 7), fill(2.0, 4), fill(1.0, 7))]
```

but ofc this is potentially more confusing — thoughts?

Reply:
Hmm, the vector isn't this code block though – it's the one above. I do see your point though! I'm sorry if I sound too conservative here about code folding :) It's mainly because I'd like to preserve a consistent user experience across the docs. But if there are good places we can apply folding to in the docs I'm super happy to start using it.

For this particular vector, would this work maybe?

```julia
y = [fill(1.0, 6)..., fill(2.0, 6)..., fill(3.0, 7)..., fill(2.0, 4)..., fill(1.0, 7)...]
```